### PR TITLE
[batch] better status information for jobs

### PR DESCRIPTION
**Pull request by danking**
1. Add `dictfix`, a little tool for wrangling json-like data into a sensible,
   standardized format.

2. Modify job.html to include a simple table of important parts of the job
   specification such as the command and the requests of cpu, memory, and
   storage.

3. Modify job.html to include a simple table summarizing the status and runtime
   of each step of the job.

My apologies for reformatting the whole file to have no excess spaces. Use the no whitespace diff option in GitHub for a better diff.

### CODE REVIEW

**Comment by danking** [comment]
example of a finished, failing job:
Screen Shot 2020-11-30 at 6 28 10 PM
and after the logs you see the spec:

Screen Shot 2020-11-30 at 6 28 23 PM

**Comment by danking** [comment]
And a fast, successful job with little output and some missing steps.
Screen Shot 2020-11-30 at 6 31 53 PM

File: batch/batch/front_end/templates/job.html [hunk 1]
```diff
@@ -1,73 +1,141 @@
...
-      <pre>{{ job_log['input'] }}</pre>
+<h2>Step Status</h2>
+<table class="data-table">
+  <thead>
+    <tr>
+	    <th>Job Step</th>
+	    <th>Image Pulling Time (s)</th>
+	    <th>Running Time (s)</th>
+	    <th>Exceeded Memory Limits?</th>
+	    <th>State</th>
```
Comment (jigold) [inline batch/batch/front_end/templates/job.html:60]: Do we need this for CI as well?

File: batch/batch/front_end/templates/job.html [hunk 2]
```diff
@@ -1,73 +1,141 @@
...
+    </tr>
+  </thead>
+  <tbody>
+    {% for step in step_statuses %}
+    {% if step %}
+    <tr>
+	    <td>{{ step['name'] }}</td>
+	    <td>
+        {% if step['timing']['pulling'] %}
```
Comment (jigold) [inline batch/batch/front_end/templates/job.html:69]: The indentation of these lines is off.

File: batch/batch/front_end/templates/job.html [hunk 3]
```diff
@@ -1,73 +1,141 @@
...
+  </thead>
+  {% if job_specification %}
+  <tbody>
+    <tr>
+      <td>image</td>
+      <td>{{ job_specification['image'] }}</td>
+    </tr>
+    <tr>
+      <td>command</td>
+      <td><pre style="max-height: 10em; max-width: 80em;">{{ "'" + (job_specification['command'] | join("' '")) + "'" }}</pre></td>
```
Comment (jigold) [inline batch/batch/front_end/templates/job.html:127]: Do we need environment variables as well for the command to be helpful to the user?

File: hail/python/test/hailtop/test_dictfix.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+def test_type_asserts():
+    actual = {'x': 1}
+    try:
+        dictfix.dictfix(actual, {'x': str})
+    except AssertionError:
+        pass
+    else:
+        assert False, actual
+
+    actual = {'x': 'hi'}
```
Comment (jigold) [inline hail/python/test/hailtop/test_dictfix.py:51]: What happens if `actual = {'x': None}`? Does this raise an exception? Does the type parameter imply non-optional values?

File: hail/python/hailtop/dictfix.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        if x is not None:
+            x = _dictfix(x, spec.subspec)
+    elif isinstance(spec, type):
+        assert isinstance(x, spec), f'{x!r} should be {spec!r}'
+    elif isinstance(spec, dict):
+        if x is None:
+            x = dict()
+        for k, subspec in spec.items():
+            v = None
+            if k in x:
```
Comment (jigold) [inline hail/python/hailtop/dictfix.py:23]: Why not just replace these three lines with `v = x.get(k)`?

**Review by jigold — CHANGES_REQUESTED** [review]

File: batch/batch/front_end/templates/job.html (hunk 3 above)
Comment (danking) [reply batch/batch/front_end/templates/job.html:127]: Screen Shot 2020-12-03 at 2 16 51 PM

**Review by jigold — APPROVED** [review]

**Pull request merged**